repose-aggregator/components/services/open-tracing-service/open-tracing-service-api/src/main/resources/META-INF/schema/config/open-tracing.xsd

```diff
@@ -0,0 +1,337 @@
+        <xs:annotation>
+            <xs:documentation>
+                <html:p>OpenTracing configuration settings.</html:p>
+                <html:p>
+                    Please refer to the
+                    <html:a href="https://github.com/opentracing/specification/blob/master/specification.md">
+                        OpenTracing specification
+                    </html:a>
+                    for more details.
+                </html:p>
+            </xs:documentation>
+        </xs:annotation>
+
+        <xs:choice>
+            <xs:element name="jaeger" type="JaegerTracerConfig"/>
+        </xs:choice>
+
+        <xs:attribute name="service-name" type="xs:string" use="required">
+            <xs:annotation>
+                <xs:documentation>
+                    <html:p>
+                        The name of the service that is reporting traces.
+                        This value will be used to identify the service and should therefore be unique
+                        (e.g., identity-repose).
+                    </html:p>
+                </xs:documentation>
+            </xs:annotation>
+        </xs:attribute>
+    </xs:complexType>
+
+    <xs:complexType name="TracerConfig" abstract="true"/>
+
+    <xs:complexType name="JaegerTracerConfig">
+        <xs:complexContent>
+            <xs:extension base="TracerConfig">
+                <xs:annotation>
+                    <xs:documentation>
+                        <html:p>
+                            Jaeger sampling configuration.
+                            As of v0.21.0, there are 4 types of configurations:
+                            - constant (either sample or not)
+                            - probabilistic (probability of sampling between 0.0 (never) and 1.0 (always)
+                            - rate limiting (maximum traces per second.  Defaults to 1.0 traces per second)
+                            - guaranteed throughput (combination of rate limiting and probabilistic) (not currently supported)
+                            - per operation (operation/specific probabilistic sampling.  Samples on names of requests) (not currently supported)
+                        </html:p>
+                    </xs:documentation>
+                </xs:annotation>
+
+                <xs:sequence>
+                    <xs:choice id="jaegerConnection">
+                        <xs:element name="connection-udp" type="JaegerConnectionUdp"/>
+                        <xs:element name="connection-http" type="JaegerConnectionHttp"/>
+                    </xs:choice>
+
+                    <xs:choice id="jaegerSampling">
+                        <xs:element name="sampling-constant" type="JaegerSamplingConstant"/>
+                        <xs:element name="sampling-probabilistic" type="JaegerSamplingProbabilistic"/>
+                        <xs:element name="sampling-rate-limiting" type="JaegerSamplingRateLimiting"/>
+                    </xs:choice>
```

Comment:
I tried three separate approached to clean this up a bit:
1. `xs:choice` with bindings. This is what you see here and in the `bindings.xjb` file. Basically, allows the user to choice one of the enumerated `element`s. The bindings magic just binds the resultant type in the generated object to the `JaegerSampling` parent type so that we have a more natural expression of polymorphism in the JVM world.
1. XML polymorphism. Look through the commit history for a commit about using raw XML types to see the implementation. Rather than giving the user a choice, force the user to always supply an `element` of the parent type (e.g., `JaegerSampling`). This cleans up both the XML (by not requiring us to enumerate the types we've already defined) and the generate Java object. However, it does require the user to be explicit about the type of their `element` by using the `xsi:type` attribute.
Honestly, this felt like the most natural approach to me, but the user experience is debatably worse than the other choices.
1. `substitutionGroup`.  This seemed like the natural XML way to handle this case, which is why I gave it a shot. I have the implementation stashed locally because JAXB does not play well with substitution groups. That is, JAXB will treat the substituted `element` as a `JAXBElement` and return that rather than a generated type, which makes manipulation of the generated object quite cumbersome. Substitution groups allow the user to choose an `element` by name rather choosing a type (similar to the `choice` approach). It also does not require binding magic. It does, however, still require enumeration of already defined types, and that enumeration takes the form of additional elements at the root level of the schema.
Here's a partial implementation for the sake of completeness:
```
<xs:complexType name="JaegerTracerConfig">
  ...
  <xs:sequence>
      <xs:element ref="connection"/>
  </xs:sequence>
  ...
</xs:complexType

<xs:element name="connection" type="JaegerConnection" abstract="true"/>
<xs:element name="connection-http" type="JaegerConnectionHttp" substitutionGroup="connection"/>
    <xs:element name="connection-udp" type="JaegerConnectionUdp" substitutionGroup="connection"/>
```

So that's the story. I am leaving this comment so that we can discuss which approach is best.

Reply:
I like Option 1. I understand that Option 2 is the more XML way to do things, but the typical Repose user/maintainer doesn't want to deal with XML any more than they have too. Option 3 leads us to a tooling issue and since JAXB is how we do things now and into the foreseeable future, I have to vote against it simply for clarity.

My 2&#162; are Option 1.